exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -95,99 +108,103 @@
+                  ServiceRuntime serviceRuntime) {
     super(nativeHandle, true);
-    populateServiceMap(serviceAdapters);
+    this.serviceRuntime = serviceRuntime;
+    timeServiceIds = timeServiceSpecs.stream()
+        .map(t -> t.serviceId)
+        .collect(toSet());
   }
 
-  private static TestKit newInstance(List<Class<? extends ServiceModule>> serviceModules,
+  private static TestKit newInstance(TestKitServiceInstances[] serviceInstances,
                                      EmulatedNodeType nodeType, short validatorCount,
-                                     @Nullable TimeProvider timeProvider) {
-    Injector frameworkInjector = Guice.createInjector(new TestKitFrameworkModule());
-    return newInstanceWithInjector(serviceModules, nodeType, validatorCount, timeProvider,
-        frameworkInjector);
-  }
-
-  private static TestKit newInstanceWithInjector(
-      List<Class<? extends ServiceModule>> serviceModules, EmulatedNodeType nodeType,
-      short validatorCount, @Nullable TimeProvider timeProvider, Injector frameworkInjector) {
-    Map<Short, UserServiceAdapter> serviceAdapters = toUserServiceAdapters(
-        serviceModules, frameworkInjector);
+                                     List<TimeServiceSpec> timeServiceSpecs,
+                                     Path artifactsDirectory, int serverPort) {
+    ServiceRuntimeAdapter serviceRuntimeAdapter =
+        getServiceRuntimeAdapter(artifactsDirectory, serverPort);
     boolean isAuditorNode = nodeType == EmulatedNodeType.AUDITOR;
-    UserServiceAdapter[] userServiceAdapters = serviceAdapters.values()
-        .toArray(new UserServiceAdapter[0]);
-    TimeProviderAdapter timeProviderAdapter =  timeProvider == null
-        ? null
-        : new TimeProviderAdapter(timeProvider);
-    long nativeHandle = nativeCreateTestKit(userServiceAdapters, isAuditorNode, validatorCount,
-        timeProviderAdapter);
-    return new TestKit(nativeHandle, serviceAdapters);
+    long nativeHandle = nativeCreateTestKit(serviceInstances, isAuditorNode, validatorCount,
+        timeServiceSpecs.toArray(new TimeServiceSpec[0]), serviceRuntimeAdapter);
+    ServiceRuntime serviceRuntime = serviceRuntimeAdapter.getServiceRuntime();
+    return new TestKit(nativeHandle, timeServiceSpecs, serviceRuntime);
   }
 
-  /**
-   * Returns a list of user service adapters created from given service modules.
-   */
-  private static Map<Short, UserServiceAdapter> toUserServiceAdapters(
-      List<Class<? extends ServiceModule>> serviceModules, Injector frameworkInjector) {
-    List<UserServiceAdapter> services = serviceModules.stream()
-        .map(s -> createUserServiceAdapter(s, frameworkInjector))
-        .collect(toList());
-    // UserServiceAdapters are removed, so this usage is a false signal for keeping getId()
-    return Maps.uniqueIndex(services, UserServiceAdapter::getId);
-  }
+  private static ServiceRuntimeAdapter getServiceRuntimeAdapter(
```

Comment:
createServiceRuntime/newServiceRuntime, as this method creates the whole thing with its dependencies each time it is invoked.